src/Text/Parsing/Parser/String.purs

```diff
@@ -2,89 +2,127 @@
 
 module Text.Parsing.Parser.String where
 
+
+import Control.Monad.Rec.Class (tailRecM3, Step(..))
 import Data.String as S
 import Control.Monad.State (modify, gets)
-import Data.Array (many)
-import Data.Foldable (elem, notElem)
+import Data.Array (many, toUnfoldable)
+import Data.Foldable (elem, notElem, foldMap)
+import Data.Unfoldable (class Unfoldable)
+import Data.List as L
 import Data.Maybe (Maybe(..))
-import Data.Newtype (wrap)
-import Data.String (Pattern, fromCharArray, length, singleton)
+import Data.Either (Either(..))
+import Data.Monoid (class Monoid)
 import Text.Parsing.Parser (ParseState(..), ParserT, fail)
 import Text.Parsing.Parser.Combinators (try, (<?>))
-import Text.Parsing.Parser.Pos (updatePosString)
+import Text.Parsing.Parser.Pos (Position, updatePosString, updatePosChar)
 import Prelude hiding (between)
+import Data.Foldable (foldl)
+
+-- | A newtype used in cases where there is a prefix string to droped.
+newtype Prefix f = Prefix f
+
+derive instance eqPrefix :: Eq f => Eq (Prefix f)
+derive instance ordPrefix :: Ord f => Ord (Prefix f)
+-- derive instance newtypePrefix :: Newtype Prefix _
+
+instance showPrefix :: Show f => Show (Prefix f) where
+  show (Prefix s) = "(Prefix " <> show s <> ")"
+
+class HasUpdatePosition a where
+  updatePos :: Position -> a -> Position
+
+instance stringHasUpdatePosition :: HasUpdatePosition String where
+  updatePos = updatePosString
+
+instance charHasUpdatePosition :: HasUpdatePosition Char where
+  updatePos = updatePosChar
 
 -- | This class exists to abstract over streams which support the string-like
 -- | operations which this modules needs.
-class StringLike s where
-  drop :: Int -> s -> s
-  indexOf :: Pattern -> s -> Maybe Int
-  null :: s -> Boolean
-  uncons :: s -> Maybe { head :: Char, tail :: s }
-
-instance stringLikeString :: StringLike String where
-  uncons = S.uncons
-  drop = S.drop
-  indexOf = S.indexOf
-  null = S.null
-
--- | Match end-of-file.
-eof :: forall s m. StringLike s => Monad m => ParserT s m Unit
+-- |
+-- | Instances must satisfy the following laws:
+-- |
+class StreamLike f c | f -> c where
+  uncons :: f -> Maybe { head :: c, tail :: f, updatePos :: (Position -> Position) }
+  drop :: Prefix f -> f -> Maybe {  rest :: f, updatePos :: (Position -> Position) }
+
+instance stringLikeString :: StreamLike String Char where
+  uncons f = S.uncons f <#> \({ head, tail}) ->
+    { head: head, updatePos: (_ `updatePos` head), tail}
+  drop (Prefix p) s = S.stripPrefix (S.Pattern p) s <#> \rest ->
+    { rest: rest, updatePos: (_ `updatePos` p)}
+
+instance listcharLikeString :: (Eq a, HasUpdatePosition a) => StreamLike (L.List a) a where
+  uncons f = L.uncons f <#> \({ head, tail}) ->
+    { head: head, updatePos: (_ `updatePos` head), tail}
+  drop (Prefix p') s' = case (tailRecM3 go p' s' id) of -- no MonadRec for Maybe
```

Comment:
I'm not sure. I'm tempted to say no, but if you'd like to open a PR, we can discuss it.